DBZ-7461 Update doc for snapshot modes alignment (#5316)
debezium/debezium: documentation/modules/ROOT/pages/connectors/postgresql.adoc

* It is known that some WAL segments have been deleted and are no longer available. +
* After a cluster failure, a new primary has been promoted. The `always` snapshot mode ensures that the connector does not miss any changes that were made after the new primary had been promoted but before the connector was restarted on the new primary.

|`initial` (default)
|The connector performs a database snapshot when no Kafka offsets topic exists. After the database snapshot completes the Kafka offsets topic is written. If there is a previously stored LSN in the Kafka offsets topic, the connector continues streaming changes from that position.

|`initial_only`
|The connector performs a database snapshot and stops before streaming any change event records. If the connector had started but did not complete a snapshot before stopping, the connector restarts the snapshot process and stops when the snapshot completes.

|`no_data`

> |The connector never performs snapshots. When a connector is configured this way, its behavior when it starts is as follows. If there is a previously stored LSN in the Kafka offsets topic, the connector continues streaming changes from that position. If no LSN has been stored, the connector starts streaming changes from the point in time when the PostgreSQL logical replication slot was created on the server. This snapshot mode is useful only when you know all data of interest is still reflected in the WAL.

Reviewer: ```suggestion
|The connector never performs snapshots. 
When a connector is configured this way, after it starts, it behaves as follows:

If there is a previously stored LSN in the Kafka offsets topic, the connector continues streaming changes from that position. 
If no LSN is stored, the connector starts streaming changes from the point at which the PostgreSQL logical replication slot was created on the server. 
Use this snapshot mode only when you know that all data of interest is still reflected in the WAL.
```

|[.line-through]`never`
|Deprecated, see `no_data`

|`when_needed`

> |The connector performs a snapshot upon startup whenever it deems it necessary. That is, when no offsets are available, or when a previously recorded offset specifies a log position that is not available in the server.

Reviewer: ```suggestion
|After the connector starts, it performs a snapshot only if it detects one of the following circumstances:
 
* It cannot detect any topic offsets.
* A previously recorded offset specifies a log position that is not available on the server.
```

ifdef::community[]
|`custom`
|The `custom` snapshot mode lets you inject your own implementation of the `io.debezium.spi.snapshot.Snapshotter` interface.

> Set the `snapshot.mode.custom.name` configuration property to the name provided by the `name()` method of your implementation, as specified on the classpath of your Kafka Connect cluster, or included in the connector JAR file, if using the `EmbeddedEngine`.

Reviewer: ```suggestion
Set the `snapshot.mode.custom.name` configuration property to the name provided by the `name()` method of your implementation. 
The name is specified on the classpath of your Kafka Connect cluster.
If you use the `EmbeddedEngine`, the name is included in the connector JAR file.
```

> For more information, see xref:connector-custom-snapshot[custom snapshotter SPI].

Reviewer: Ditto.

include::{partialsdir}/modules/all-connectors/proc-stopping-an-incremental-snapshot-kafka.adoc[leveloffset=+1]

ifdef::community[]

> [[connector-custom-snapshot]]

Reviewer: Ditto.

 isbn.schema.name: io.debezium.postgresql.type.Isbn

|[[postgresql-property-snapshot-mode]]<<postgresql-property-snapshot-mode, `+snapshot.mode+`>>
|_initial_
|Specifies the criteria for performing a snapshot when the connector starts: +

`always`:: The connector performs a snapshot each time the connector starts.

`initial`:: The connector performs a snapshot only when no offsets have been recorded for the logical server name.

`initial_only`:: The connector performs an initial snapshot and then stops, without processing any subsequent changes.

> `no_data`:: The connector never performs snapshots. When a connector is configured this way, its behavior when it starts is as follows. If there is a previously stored LSN in the Kafka offsets topic, the connector continues streaming changes from that position. If no LSN has been stored, the connector starts streaming changes from the point in time when the PostgreSQL logical replication slot was created on the server. This snapshot mode is useful only when you know all data of interest is still reflected in the WAL.

Reviewer: ```suggestion
`no_data`:: The connector never performs snapshots. 
When a connector is configured this way, after it starts, it behaves as follows:

If there is a previously stored LSN in the Kafka offsets topic, the connector continues streaming changes from that position.
If no LSN is stored, the connector starts streaming changes from the point in time when the PostgreSQL logical replication slot was created on the server. 
Use this snapshot mode only when you know all data of interest is still reflected in the WAL.
```

`never`:: Deprecated see `no_data`.

> `when_needed`:: The connector performs a snapshot upon startup whenever it deems it necessary. That is, when no offsets are available, or when a previously recorded offset specifies a log position that is not available in the server.

Reviewer: ```suggestion
`when_needed`:: After the connector starts, it performs a snapshot only if it detects one of the following circumstances:
 
* It cannot detect any topic offsets.
* A previously recorded offset specifies a log position that is not available on the server.
```

ifdef::community[]
`custom`:: The connector performs a snapshot according to the implementation specified by the xref:postgresql-property-snapshot-mode-custom-name[`snapshot.mode.custom.name`] property, which defines a custom implementation of the `io.debezium.spi.snapshot.Snapshotter` interface.
endif::community[]

For more information, see the xref:#postgresql-connector-snapshot-mode-options[table of `snapshot.mode` options].

ifdef::community[]
|[[postgresql-property-snapshot-mode-custom-name]]<<postgresql-property-snapshot-mode-custom-name, `+snapshot.mode.custom.name+`>>
|No default
| When `snapshot.mode` is set as `custom`, use this setting to specify the name of the custom implementation provided in the `name()` method that is defined by the 'io.debezium.spi.snapshot.Snapshotter' interface.
The provided implementation is called after a connector restart to determine whether to perform a snapshot.

> For more information, see xref:connector-custom-snapshot[custom snapshotter SPI].

Reviewer: Similar comment about ensuring unique IDs if content becomes productized.

 isbn.schema.name: io.debezium.postgresql.type.Isbn

|[[postgresql-property-snapshot-mode]]<<postgresql-property-snapshot-mode, `+snapshot.mode+`>>
|_initial_
|Specifies the criteria for performing a snapshot when the connector starts: +

> `always`:: The connector performs a snapshot each time the connector starts.

Reviewer: ```suggestion
`always`:: The connector performs a snapshot every time that it starts.
The snapshot includes the structure and data of the captured tables.
Specify this value to populate topics with a complete representation of the data from the captured tables every time that the connector starts.
After the snapshot completes, the connector begins to stream event records for subsequent database changes.
```

|[[postgresql-property-snapshot-locking-mode-custom-name]]<<postgresql-property-snapshot-locking-mode-custom-name, `+snapshot.locking.mode.custom.name+`>>
|No default

> | When `snapshot.locking.mode` is set as `custom`, use this setting to specify the name of the custom implementation provided in the `name()` method that is defined by the 'io.debezium.spi.snapshot.SnapshotLock' interface.

Reviewer: ```suggestion
| When xref:postgresql-property-snapshot-locking-mode[`snapshot.locking.mode`] is set to `custom`, use this setting to specify the name of the custom implementation provided in the `name()` method that is defined by the 'io.debezium.spi.snapshot.SnapshotLock' interface.
```

> For more information, see xref:connector-custom-snapshot[custom snapshotter SPI].

Reviewer: Ditto.

|Specifies how the connector queries data while performing a snapshot. +
Set one of the following options:

> `select_all`:: `select_all`:: The connector performs a `select all` query by default, optionally adjusting the columns selected based on the column include and exclude list configurations.

Reviewer 2: ```suggestion
`select_all`:: The connector performs a `select all` query by default, optionally adjusting the columns selected based on the column include and exclude list configurations.
```
Same?

|[[postgresql-property-snapshot-snapshot-query-mode-custom-name]]<<postgresql-property-snapshot-query-mode-custom-name, `+snapshot.query.mode.custom.name+`>>
|No default
| When xref:postgresql-property-snapshot-query-mode[`snapshot.query.mode`] is set as `custom`, use this setting to specify the name of the custom implementation provided in the `name()` method that is defined by the 'io.debezium.spi.snapshot.SnapshotQuery' interface.

> For more information, see xref:connector-custom-snapshot[custom snapshotter SPI].

Reviewer: Ditto.
